### PR TITLE
fs: add littlefs as a package

### LINKED ISSUE #8204: pkg: port LittleFS
Offline @benpicco told me very excited how much faster and smaller than SPIFFS [LittleFS](https://github.com/geky/littlefs) is. A quick glance at the project's README let's me think that this might be a nice file system to support for RIOT. Anyone wants to port it to RIOT with a `vfs` wrapper?

### PR DESCRIPTION
**Pull request by vincent-d**
Just to avoid duplicate work, I'm pushing this. I started to work on littlefs package integration. I wanted to test it and see the performances.
I still have some issues and I don't know when I'll have time to continue working on it.

Fix #8204

### CODE REVIEW

**Comment by vincent-d** [comment]
Looks to work now (at least unittests do not crash anymore ;)
If someone wants to play with it, it's possible as is. It would need some more clean up if we want to integrate it.

I found out that littlefs uses `malloc`, there is an option to not use it, but then only one file can be opened at a time...

**Comment by vincent-d** [comment]
Removed the WIP. Should be ready for review.

File: pkg/littlefs/fs/littlefs_fs.c [hunk 1]
```diff
@@ -0,0 +1,499 @@
...
+        return -ENOMEM;
+    default:
+        return err;
+    }
+}
+
+static int _dev_read(const struct lfs_config *c, lfs_block_t block,
+                 lfs_off_t off, void *buffer, lfs_size_t size)
+{
+    (void)block;
```
Comment (smlng) [inline pkg/littlefs/fs/littlefs_fs.c]: why, its used in DEBUG and mtd_read

File: pkg/littlefs/fs/littlefs_fs.c [hunk 2]
```diff
@@ -0,0 +1,499 @@
...
+    }
+    else {
+        return ret;
+    }
+}
+
+static int _dev_write(const struct lfs_config *c, lfs_block_t block,
+                  lfs_off_t off, const void *buffer, lfs_size_t size)
+{
+    (void)block;
```
Comment (smlng) [inline pkg/littlefs/fs/littlefs_fs.c]: dito

File: pkg/littlefs/fs/littlefs_fs.c [hunk 3]
```diff
@@ -0,0 +1,499 @@
...
+{
+    mtd_dev_t *mtd = c->context;
+
+    DEBUG("lfs_erase: c=%p, block=%" PRIu32 "\n", (void *)c, block);
+
+    int ret = mtd_erase(mtd, block * c->block_size, c->block_size);
+    if (ret >= 0) {
+        return 0;
+    }
+    else {
```
Comment (smlng) [inline pkg/littlefs/fs/littlefs_fs.c]: else not needed, due to return above

File: sys/include/fs/littlefs_fs.h [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+#if VFS_DIR_BUFFER_SIZE < 44
+#error "VFS_DIR_BUFFER_SIZE is too small, at least 44 bytes is required"
+#endif
+
+/**
+ * @name    littlefs configuration
+ * @{
+ */
+#ifndef LITTLEFS_LOOKAHED_SIZE
+/** Default lookahed size */
```
Comment (smlng) [inline sys/include/fs/littlefs_fs.h]: typo -> `look ahead`?

File: cpu/native/mtd/mtd_native.c [hunk 5]
```diff
@@ -37,7 +37,7 @@ static int _init(mtd_dev_t *dev)
     FILE *f = real_fopen(_dev->fname, "r");
 
     if (!f) {
-        DEBUG("mtd_native: init: creating file %s\n", name);
+        DEBUG("mtd_native: init: creating file %s\n", _dev->fname);
```
Comment (smlng) [inline cpu/native/mtd/mtd_native.c]: nice catch!

File: pkg/littlefs/Makefile.littlefs [hunk 6]
```diff
@@ -0,0 +1,5 @@
+MODULE := littlefs
+
+CFLAGS += -Wno-unused-parameter -Wno-format
```
Comment (smlng) [inline pkg/littlefs/Makefile.littlefs]: mhm, might be worth upstreaming some fixing then 😉

File: pkg/littlefs/fs/littlefs_fs.c [hunk 7]
```diff
@@ -0,0 +1,499 @@
+/*
+ * Copyright (C) 2017 OTA keys S.A.
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     fs
```
Comment (smlng) [inline pkg/littlefs/fs/littlefs_fs.c]: should be `pkg_littlefs`?  Btw. that group is not defined, as far as I can see, right? should add a `doc.txt` to do so.

File: pkg/littlefs/fs/littlefs_fs.c [hunk 8]
```diff
@@ -0,0 +1,499 @@
...
+    mtd_dev_t *mtd = c->context;
+
+    DEBUG("lfs_read: c=%p, block=%" PRIu32 ", off=%" PRIu32 ", buf=%p, size=%" PRIu32 "\n",
+          (void *)c, block, off, buffer, size);
+
+    int ret = mtd_read(mtd, buffer, block * c->block_size + off, size);
+    if (ret >= 0) {
+        return 0;
+    }
+    else {
```
Comment (smlng) [inline pkg/littlefs/fs/littlefs_fs.c]: else not needed, due to return above

File: pkg/littlefs/fs/littlefs_fs.c [hunk 9]
```diff
@@ -0,0 +1,499 @@
...
+    mtd_dev_t *mtd = c->context;
+
+    DEBUG("lfs_write: c=%p, block=%" PRIu32 ", off=%" PRIu32 ", buf=%p, size=%" PRIu32 "\n",
+          (void *)c, block, off, buffer, size);
+
+    int ret = mtd_write(mtd, buffer, block * c->block_size + off, size);
+    if (ret >= 0) {
+        return 0;
+    }
+    else {
```
Comment (smlng) [inline pkg/littlefs/fs/littlefs_fs.c]: else not needed

File: tests/unittests/tests-littlefs/tests-littlefs.c [hunk 10]
```diff
@@ -0,0 +1,403 @@
...
+    res = vfs_write(fd, buf, sizeof(buf));
+    TEST_ASSERT_EQUAL_INT(sizeof(buf), res);
+
+    res = vfs_close(fd);
+    TEST_ASSERT_EQUAL_INT(0, res);
+
+    res = vfs_unlink("/test-littlefs/test.txt");
+    TEST_ASSERT_EQUAL_INT(0, res);
+}
+#include <stdio.h>
```
Comment (smlng) [inline tests/unittests/tests-littlefs/tests-littlefs.c]: why here?

File: tests/unittests/tests-littlefs/tests-littlefs.c [hunk 11]
```diff
@@ -0,0 +1,403 @@
+/*
+ * Copyright (C) 2016 OTA keys S.A.
```
Comment (smlng) [inline tests/unittests/tests-littlefs/tests-littlefs.c]: 2018 is close, anyway 2017 is the year?

File: sys/include/fs/littlefs_fs.h [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @defgroup    sys_littlefs  littlefs integration
+ * @ingroup     pkg_littlefs
+ * @brief       RIOT integration of littlefs
+ *
+ * Integration of littlefs into RIOT
```
Comment (smlng) [inline sys/include/fs/littlefs_fs.h]: sounds like what `@brief` is already telling 😉

File: cpu/native/mtd/mtd_native.c (hunk 5 above)
Comment (vincent-d) [reply cpu/native/mtd/mtd_native.c]: The credit goes to @MichelRottleuthner (this comes from #7104) ;)

Comment (smlng) [reply cpu/native/mtd/mtd_native.c]: ah, so this is kind of based on #7104? and we should merge that first?

File: sys/include/fs/littlefs_fs.h (hunk 4 above)
Comment (vincent-d) [reply sys/include/fs/littlefs_fs.h]: Fixed, here and in variable names

File: tests/unittests/tests-littlefs/tests-littlefs.c (hunk 10 above)
Comment (vincent-d) [reply tests/unittests/tests-littlefs/tests-littlefs.c]: Debug leftover, removed

File: sys/include/fs/littlefs_fs.h (hunk 12 above)
Comment (vincent-d) [reply sys/include/fs/littlefs_fs.h]: Removed

File: pkg/littlefs/fs/littlefs_fs.c (hunk 3 above)
Comment (vincent-d) [reply pkg/littlefs/fs/littlefs_fs.c]: Removed

File: pkg/littlefs/fs/littlefs_fs.c (hunk 9 above)
Comment (vincent-d) [reply pkg/littlefs/fs/littlefs_fs.c]: Removed

File: pkg/littlefs/fs/littlefs_fs.c (hunk 2 above)
Comment (vincent-d) [reply pkg/littlefs/fs/littlefs_fs.c]: Removed

File: pkg/littlefs/fs/littlefs_fs.c (hunk 8 above)
Comment (vincent-d) [reply pkg/littlefs/fs/littlefs_fs.c]: Removed

File: pkg/littlefs/fs/littlefs_fs.c (hunk 1 above)
Comment (vincent-d) [reply pkg/littlefs/fs/littlefs_fs.c]: Removed

File: pkg/littlefs/fs/littlefs_fs.c (hunk 7 above)
Comment (vincent-d) [reply pkg/littlefs/fs/littlefs_fs.c]: True, but actually C files are not parsed by Doxygen.
There is already a `doc.txt` in the pkg directory defining the groups.

File: cpu/native/mtd/mtd_native.c (hunk 5 above)
Comment (vincent-d) [reply cpu/native/mtd/mtd_native.c]: This is not really based on #7104, I simply copied this fix and the `vfs.h` part as well.
So we can merge independently, I guess only a rebase will be needed.

File: pkg/littlefs/Makefile.littlefs (hunk 6 above)
Comment (vincent-d) [reply pkg/littlefs/Makefile.littlefs]: Yeah, we'll see later ^^

File: tests/unittests/tests-littlefs/tests-littlefs.c (hunk 11 above)
Comment (vincent-d) [reply tests/unittests/tests-littlefs/tests-littlefs.c]: oops, copy paste

**Comment by vincent-d** [comment]
Updated to latest upstream commit, and fixed unittests.

I had to patch for mips to build, so I use my fork and I opened an upstream PR (https://github.com/geky/littlefs/pull/3) to fix the errors.

**Comment by smlng** [comment]
I still like to get #7104 merged first (for the fix as discussed above) and rebase this afterwards. Otherwise this needs some reasonable squashing, too.

**Comment by smlng** [comment]
#7104 was merged, please rebase and squash

**Comment by vincent-d** [comment]
Rebased and squashed!

**Comment by smlng** [comment]
Compiled and ran unittest on macOS: works. But with DEBUG output on, I get several errors:

```
/RIOT/pkg/littlefs/fs/littlefs_fs.c:313:50: error: format specifies type 'unsigned int' but the argument has type
      'size_t' (aka 'unsigned long') [-Werror,-Wformat]
          (void *)filp, (void *)fp, (void *)src, nbytes);
                                                 ^~~~~~
/RIOT/core/include/debug.h:95:32: note: expanded from macro 'DEBUG'
#define DEBUG(...) DEBUG_PRINT(__VA_ARGS__)
                               ^~~~~~~~~~~
/RIOT/core/include/debug.h:56:20: note: expanded from macro 'DEBUG_PRINT'
            printf(__VA_ARGS__); \
                   ^~~~~~~~~~~
/RIOT/pkg/littlefs/fs/littlefs_fs.c:329:51: error: format specifies type 'unsigned int' but the argument has type
      'size_t' (aka 'unsigned long') [-Werror,-Wformat]
          (void *)filp, (void *)fp, (void *)dest, nbytes);
                                                  ^~~~~~
/RIOT/core/include/debug.h:95:32: note: expanded from macro 'DEBUG'
#define DEBUG(...) DEBUG_PRINT(__VA_ARGS__)
                               ^~~~~~~~~~~
/RIOT/core/include/debug.h:56:20: note: expanded from macro 'DEBUG_PRINT'
            printf(__VA_ARGS__); \
                   ^~~~~~~~~~~
/RIOT/pkg/littlefs/fs/littlefs_fs.c:345:37: error: format specifies type 'long' but the argument has type 'off_t'
      (aka 'long long') [-Werror,-Wformat]
          (void *)filp, (void *)fp, off, whence);
                                    ^~~
/RIOT/core/include/debug.h:95:32: note: expanded from macro 'DEBUG'
#define DEBUG(...) DEBUG_PRINT(__VA_ARGS__)
                               ^~~~~~~~~~~
/RIOT/core/include/debug.h:56:20: note: expanded from macro 'DEBUG_PRINT'
            printf(__VA_ARGS__); \
                   ^~~~~~~~~~~
/RIOT/pkg/littlefs/fs/littlefs_fs.c:360:19: error: use of undeclared identifier 'filp'
          (void *)filp, (void *)buf);
                  ^
/RIOT/pkg/littlefs/fs/littlefs_fs.c:396:19: error: use of undeclared identifier 'filp'
          (void *)filp, (void *)buf);
                  ^
5 errors generated.

```

**Comment by vincent-d** [comment]
Oops, will fix

**Comment by vincent-d** [comment]
Should be fixed. Amended directly.

I don't have the first errors on my machine, so it would be great if you could confirm it's fixed.

**Comment by smlng** [comment]
re-ACK, works like charm with and w/o Debug on macOS! NICE!

**Comment by smlng** [comment]
then let's do this: ACK & GO

File: pkg/littlefs/fs/littlefs_fs.c [hunk 13]
```diff
@@ -0,0 +1,497 @@
...
+
+    DEBUG("littlefs: readdir: dirp=%p, entry=%p\n",
+          (void *)dirp, (void *)entry);
+
+    struct lfs_info info;
+    int ret = lfs_dir_read(&fs->fs, dir, &info);
+    if (ret >= 0) {
+        entry->d_ino = info.type;
+        entry->d_name[0] = '/';
+        strncpy(entry->d_name + 1, info.name, VFS_NAME_MAX - 1);
```
Comment (benpicco) [inline pkg/littlefs/fs/littlefs_fs.c:445]: Why is a `/` added in front of every directory entry here?

**Pull request merged**